core/src/main/java/org/apache/iceberg/ManifestFilterManager.java

```diff
@@ -510,16 +510,19 @@ private Pair<InclusiveMetricsEvaluator, StrictMetricsEvaluator> metricsEvaluator
       // in other words, ResidualEvaluator returns a part of the expression that needs to be
       // evaluated
       // for rows in the given partition using metrics
-      return metricsEvaluators.computeIfAbsent(
-          file.partition(),
-          partition -> {
-            Expression residual = residualEvaluator.residualFor(partition);
-            InclusiveMetricsEvaluator inclusive =
-                new InclusiveMetricsEvaluator(tableSchema, residual, caseSensitive);
-            StrictMetricsEvaluator strict =
-                new StrictMetricsEvaluator(tableSchema, residual, caseSensitive);
-            return Pair.of(inclusive, strict);
-          });
+      PartitionData partition = (PartitionData) file.partition();
```

Comment:
Should we use `StructLikeMap` here instead of relying on `PartitionData` in a regular map? We could use a pattern like in `FanoutWriter` to keep these by partition spec ID.

Reply:
It is a StructLikeMap, this actually causes the collisions to happen more frequently because of the structlikewrapper hashcode function

Reply:
Checking fanout writer, it actually does the same pattern I implemented here, but uses StructCopy.copy instead of the PartitionData.copy method. That may be a safer thing to do since technically we are not guaranteed the PartitionData type.

Reply:
Using StructCopy is actually going to be a bit of a bigger change if we want to keep the interface package private since we need to move it into the `org.apache.iceberg` package.

Reply:
Ok I found a bug in StructCopy as well, it doesn't copy FixedByte types and just copies array values so we actually probably need to file a issue for that.

It should be doing this

From PartitionData.copyData
```java
          case FIXED:
            byte[] buffer = (byte[]) data[i];
            copy[i] = Arrays.copyOf(buffer, buffer.length);
            break;
 ```

but instead the StructCopy code always does this
```java
      if (value instanceof StructLike) {
        values[i] = copy((StructLike) value);
      } else {
        values[i] = value;
      }
```

Reply:
Filed #6685 to track StructCopy bug